### PR TITLE
Dice genetic

### PR DESCRIPTION
**Pull request by soundarya98**
- Shifted generic code to explainer_base
- Removed tf specific computations in loss computation
- Added support for label encoding
- Obtained diverse counterfactuals through genetic algorithm
- Made dice_KD work without tensorflow models
- Updated documentation for dice_genetic in DiCE_model_agnostic_CFs.ipynb

### CODE REVIEW

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+    def compute_yloss(self, cfs):
+        """Computes the first part (y-loss) of the loss function."""
+        yloss = 0.0
+        for i in range(self.total_CFs):
+            if self.yloss_type == "l2_loss":
+                temp_loss = pow((self.model.get_output(cfs[i]) - self.target_cf_class), 2)[0][0]
+
+            elif self.yloss_type == "log_loss":
+                temp_logits = math.log((abs(self.model.get_output(cfs[i]) - 0.000001))/(1 - abs(self.model.get_output(self.cfs[i]) - 0.000001)))
+                temp_loss = self.target_cf_class[0][0] * (-1) * np.log(sigmoid(temp_logits)) + (1 - self.target_cf_class[0][0]) * (-1) * np.log(1 - sigmoid(temp_logits))
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_genetic.py:93]: where is `sigmoid` defined? your tests are failing because flake8 says sigmoid is not defined.

File: dice_ml/data_interfaces/private_data_interface.py [hunk 2]
```diff
@@ -81,6 +83,14 @@ def __init__(self, params):
             if feature_name not in self.type_and_precision:
                 self.type_and_precision[feature_name] = 'int'
 
+        self.labelencoder = {}
```
Comment (amit-sharma) [inline dice_ml/data_interfaces/private_data_interface.py:86]: can you explain in an inline code comment why we need a label encoder?

File: dice_ml/data_interfaces/private_data_interface.py [hunk 3]
```diff
@@ -182,6 +192,12 @@ def get_indexes_of_features_to_vary(self, features_to_vary='all'):
                     ixs.append(colidx)
             return ixs
 
+    def from_label(self, data):
```
Comment (amit-sharma) [inline dice_ml/data_interfaces/private_data_interface.py:195]: add a string documentation for this function just like from_dummies function below.

That holds for every new function that you add.

File: dice_ml/data_interfaces/public_data_interface.py [hunk 4]
```diff
@@ -263,6 +272,12 @@ def get_indexes_of_features_to_vary(self, features_to_vary='all'):
                     ixs.append(colidx)
             return ixs
 
+    def from_label(self, data):
```
Comment (amit-sharma) [inline dice_ml/data_interfaces/public_data_interface.py:275]: again, need some documentation string for this method.

File: dice_ml/diverse_counterfactuals.py [hunk 5]
```diff
@@ -6,7 +6,7 @@
 class CounterfactualExamples:
     """A class to store and visualize the resulting counterfactual explanations."""
 
-    def __init__(self, data_interface, test_instance, test_pred, final_cfs, final_cfs_preds, final_cfs_sparse=None, cfs_preds_sparse=None, posthoc_sparsity_param=0, desired_class="opposite"):
+    def __init__(self, data_interface, test_instance, test_pred, final_cfs, final_cfs_preds, final_cfs_sparse=None, cfs_preds_sparse=None, posthoc_sparsity_param=0, desired_class="opposite", encoding='one-hot'):
```
Comment (amit-sharma) [inline dice_ml/diverse_counterfactuals.py:9]: need to think more about the encoding parameter. For a user, it can be a problem if their ML model uses one kind of encoding, and then DiCE forces on them to choose a different kind of encoding (e.g., onehotencoder from sklearn or get_dummies from pandas). 
The same issue can also come with labelencoder. It will be nice to support any encoder that a user provides. 
To do that, we can say that the encoding parameter can be either a string or a function. If it's a string, we apply the default encoding. If it is a function, then we directly apply that function to the input data.

**Review by amit-sharma — COMMENTED** [review]
I reviewed everything except dice_genetic.py, which I will do in the next PR, as discussed.

**Review by amit-sharma — APPROVED** [review]
Looks good for a initial merge as a work in progress (WIP)
